tools/flat-storage/src/flat.rs

```diff
@@ -0,0 +1,272 @@
+                    verify_cmd.shard_id
+                ));
+                let block_header = chain_store.get_block_header(&head_hash).unwrap();
+                let shard_layout = hot_runtime.get_shard_layout(block_header.epoch_id()).unwrap();
+
+                println!(
+                    "Verifying flat storage for shard {:?} - flat head @{:?} ({:?})",
+                    verify_cmd.shard_id,
+                    block_header.height(),
+                    block_header.hash()
+                );
+                let chunk_extra = chain_store
+                    .get_chunk_extra(
+                        &head_hash,
+                        &ShardUId::from_shard_id_and_layout(verify_cmd.shard_id, &shard_layout),
+                    )
+                    .unwrap();
+
+                // The state root must be from AFTER applying the final block (that's why we're taking it from the chunk extra).
+                let state_root = chunk_extra.state_root();
+
+                println!("Verifying using the {:?} as state_root", state_root);
+                let tip = chain_store.final_head().unwrap();
+
+                hot_runtime.create_flat_storage_state_for_shard(
+                    verify_cmd.shard_id,
+                    tip.height,
+                    &chain_store,
+                );
+
+                let trie = hot_runtime
+                    .get_view_trie_for_shard(verify_cmd.shard_id, &head_hash, state_root.clone())
+                    .unwrap();
+
+                let all_entries = store_helper::iter_flat_state_entries(
+                    shard_layout,
+                    verify_cmd.shard_id,
+                    &hot_store,
+                    None,
+                    None,
+                );
+
+                let trie_iter = trie.iter().unwrap().filter(|entry| {
+                    let result_copy = &entry.clone().unwrap().0;
+                    let result = &result_copy[..];
+                    parse_account_id_from_raw_key(result).unwrap().is_some()
+                });
+
+                let mut verified = 0;
+                let mut success = true;
+                for (item_trie, item_flat) in tqdm(std::iter::zip(trie_iter, all_entries)) {
+                    let value_ref = ValueRef::decode((*item_flat.1).try_into().unwrap());
+                    verified += 1;
+
+                    let item_trie = item_trie.unwrap();
+                    if item_trie.0 != *item_flat.0 {
+                        println!(
+                            "Different keys {:?} in trie, {:?} in flat. ",
+                            item_trie.0, item_flat.0
+                        );
+                        success = false;
+                        break;
+                    }
+                    if item_trie.1.len() != value_ref.length as usize {
+                        println!(
+                            "Wrong values for key: {:?} length trie: {:?} vs {:?}",
```

Comment:
```suggestion
                            "Wrong `ValueRef::length` for key: {:?} in trie: {:?} vs. flat storage: {:?}",
```

I couldn't fully understand how to parse it, so I'm leaving my suggestion here and below